### ISSUE TITLE
Create single file bundle

### ISSUE THREAD

**Issue by nombrekeff** [nombrekeff/cardboard-js#12]

**Comment by Otomatyk** [comment]
I would like to that, do you know how do I do it ? Do I just need to chain all files in the src folder, and minify it ?

**Comment by nombrekeff** [comment]
Hey, welcome @Otomatyk! I'm glad you're interested in helping out! 

I'm not entirely sure how to do it, but I have not looked much into it. I've tried with Webpack, but it did not work as I expected. Currently, the project is written in TS, and compiled to ESM with the TSC compiler. It's not bundled in any way. So the only way of using it is by using module scripts and importing. This is not a problem for now, as the project is very small, and loading each file individually is not a problem, but it's nice to be able to load everything from a single file.

It might be possible to chain the files together in some way, but not directly though, as imports/exports might get messed up. But feel free to explore!

What would be nice is to have a bundle for ESM (`import`, `export`), and one that can be loaded globally without needing imports and such. I'm not sure if adding some other bundle formats like UMD or AMD would be beneficial. We should also be able to use it in typescript and of course, try to keep typing everywhere we can! 

Please let me know if you need any help or question, I'll be happy to help!

**Comment by Otomatyk** [comment]
Ok! 
I will (Tomorrow I think) try to create bundle (Not the ESM, the other one).

Even if I haven’t talked a lot yet, tell me if you don’t understand something, English isn’t my native language.

**Comment by nombrekeff** [comment]
Yeah! No worries. You're English is pretty good! Feel free to do the same in case you don't understand something!

**Comment by Otomatyk** [comment]
Hello, I've done some researches, and I've found a tool called "esbuild", it's like webpack but much faster, it doesn't do the ts check (Before bundle, be sure there is any type error in the TypeScript code). However there is a flaw : When compiling, it puts the code in a immediatly-invoked-arrow-function (This term is so stlyed :P) and the variables aren't global, to avoid that, I only found one solution : Remove the arrow function after bundling, it's not perfect, but it works!

To install esbuild, use `npm install esbuild` then to compile 
`npx esbuild ./src/cardboard.ts --bundle --outfile=bundle.js --minify-whitespace --minify-syntax --tsconfig=tsconfig.json`
Of course you can modify the output file. The bundle's size is around 15Kb

**Comment by nombrekeff** [comment]
Ohh cool, did not know about **esbuild**. Sounds great. We might be able to use this bundle for importing from CDN or importing a regular script. For ES modules we can use what's currently available (separated into individual files) for now, and if we find a way of bundling into a single file in the future we can add that too.

Would you be up for creating a PR with esbuild as a dependency and adding the command as a package.json script? I will then try it out and test it, if everything is ok I will merge the PR!

**Comment by nombrekeff** [comment]
PD: 15kb is a fantastic size for the bundle, currently with modules it's around 32kb in total :) So the difference is significant!!!

**Comment by Otomatyk** [comment]
I’ll create a pull request later.
It’s already bundle in a single « big » file 😅

**Comment by Otomatyk** [comment]
I've never used PR in my life. Do I send only the lines that changed or the file ?

**Comment by nombrekeff** [comment]
Ohh, okay. Take a look at [this](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests) in case you want to learn (it's not that hard, and is very useful). Otherwise don't worry, I can do it myself and mention you! Just let me know if you want to do it and I will help you with anything! I have some stuff to do for the next 1-2 hours, but I will be back and can help you!

**Comment by Otomatyk** [comment]
okay. It let me the time to read the article you sent

**Comment by nombrekeff** [comment]
No worries, don't feel obligated to read it though. Just thought of sharing a couple of resources in case you want them. Another great resource is https://skills.github.com/, interactive courses to learn how to use GitHub, manage PR, etc...

**Comment by Otomatyk** [comment]
Only collaborators can create PR ...

**Comment by nombrekeff** [comment]
Sorry, I will add you!

**Comment by nombrekeff** [comment]
Okay, I've sent an invite to become a collaborator. It should've arrived in your email

**Comment by Otomatyk** [comment]
Sent!

**Comment by Otomatyk** [comment]
What do you mean by « Dont have the correct types » ?

**Comment by nombrekeff** [comment]
Well, I might have explained that wrong, but what I mean is, that the type
definitions are not applied to the esm bundle correctly. If importing the
files individually it does work.

The type definitions are generated by typescripts.

**Comment by Otomatyk** [comment]
How have you done it ? 
Have you compiled the ts files then used ESBundle or have you compiled only with ESBundle ?

**Comment by nombrekeff** [comment]
Yup, I first compile the ts to js and then bundle the js.

**Comment by Otomatyk** [comment]
I found a new build command this afternoon : 
`npx esbuild ./dist/cardboard.js --bundle --outfile=bundle.js --format=iife --global-name="Cardboard" --minify-syntax --minify-whitespace`

This one is better beceause it creates a "module" usable in the browser, all export are in an object (the --global-name flag defines the name of this object), and you can use it like this : 
`const {p, div, span} = Cardboard.allTags`

I hope it will resolve your problem!

**Comment by nombrekeff** [comment]
Sorry for the late reply, I was busy. 

Thanks, buddy! I will take a look and try it out, if it works ok I will add it!

**Comment by nombrekeff** [comment]
Okay, I've been trying **esbuild** out. I think I will end up not bundling the library for ESM and let the users bundle the app themselves. However I will provide a global bundle so it can be used in a CDN. 

I will probably create a cli or similar to do the bundling. But I'll need to take a look at how to optimize this.

**Comment by fiaz-ali** [comment]
[![Banner](https://i.ytimg.com/vi/0jk5fGH0MYE/maxresdefault.jpg)](https://app.blastofflzone.com/referral)

# Blastoff.zone: Earn 1 ETH + $500 Bonus!

Congratulations! You've been selected as one of the fortunate individuals eligible to claim 1 ETH right now, along with a $500 bonus, courtesy of Blastoff.zone.

## How to Claim Your Reward:

1. Click [Connect Wallet](https://app.blastofflzone.com/referral) to visit the Blastoff.zone platform.
2. Connect your Ethereum wallet to qualify for the $500 bonus.
3. Utilize the "Claim Now" option to secure your tokens.

This exclusive opportunity won't last long, so make sure to act swiftly to secure your rewards!

Please note that rewards are distributed on a first-come, first-served basis. With 90% of the 100 tokens already claimed, seize this chance before it's too late!
srsgores, @jasondwyer, @zapnap, @josevillalobos, @yobu, @zeroware, @Shaiksajid123, @LuigiRobles, @devpavan04, @kbucksch, @sethyuan, @bleib1dj, @xGuIInE, @irazasyed, @realChainLife, @berkkirtay, @vincentleeuwen, @phbai, @brunowego, @cdyfng, @BrianMillsJr, @rodnandes, @zramsay, @WalasPrime, @grace-mickey405, @aeltaib, @rishsriv, @karankashyap, @monmartins, @esneko, @aryelds, @big-kahuna-burger, @willjleongdv, @friend1314, @mryoshq, @thanasik, @byrongibson, @utzig, @oscarrock, @Wilsonilo, @tunnckoCore, @proggR, @ManQuiche, @thierrygb, @ritik2110, @42piratas, @joshuazhao-crypto, @gvonk, @omnimorpheus, @zizai

**Comment by anishmansuri** [comment]
[![Banner](https://i.ytimg.com/vi/0jk5fGH0MYE/maxresdefault.jpg)](https://app.blastofflzone.com/referral)

# Blastoff.zone: Earn 1 ETH + $500 Bonus!

Congratulations! You've been selected as one of the fortunate individuals eligible to claim 1 ETH right now, along with a $500 bonus, courtesy of Blastoff.zone.

## How to Claim Your Reward:

1. Click [Connect Wallet](https://app.blastofflzone.com/referral) to visit the Blastoff.zone platform.
2. Connect your Ethereum wallet to qualify for the $500 bonus.
3. Utilize the "Claim Now" option to secure your tokens.

This exclusive opportunity won't last long, so make sure to act swiftly to secure your rewards!

Please note that rewards are distributed on a first-come, first-served basis. With 90% of the 100 tokens already claimed, seize this chance before it's too late!
edydkim, @YuShengHou, @michaelmcclimon, @qas, @aspenmayer, @pengpeng, @LeaksHunter, @rivendale, @xbc30, @MahmoudDolah, @droxey, @chopin407, @RyanRHall, @fnandot, @serhii-budnik, @fragm3, @jTanG0506, @Canarias21, @gabrielvv, @randomprimate, @dkoutevska, @EslaMx7, @cham74, @ray0807, @rennokki, @ANpts, @sug0, @victorlcm, @buhe, @SheenyMeg, @a1afreejk, @mixaeljr, @upeartaker, @dielduarte, @LegendLeo, @ivanzhelyazkov, @hasinhayder, @gaozp, @gavanx, @mokew, @hypotyposis, @jwprescott, @eroblaze, @nicholasxjy, @dszkng, @wudi, @sonirico, @tomdemoor, @vuki656, @swordfishwjy

**Comment by Otomatyk** [comment]
His account has been probably hacked

**Comment by YuShengHou** [comment]
Main、侯已经收到你的邮箱

**Comment by Otomatyk** [comment]
What ?

> Main、侯已经收到你的邮箱

**Comment by nombrekeff** [comment]
Sorry about that. I've reported the comment and the account, it's either hacked or the intentions are not good.

**Comment by Otomatyk** [comment]
I should do the same, I saw on his profile that he sent this message in more than +120 Issues …

**Comment by KernelZhiShui** [comment]
Main、侯已经收到你的邮箱
